ssl/record/rec_layer_s3.c

```diff
@@ -1554,30 +1554,44 @@ int ssl3_read_bytes(SSL *s, int type, int *recvd_type, unsigned char *buf,
         return -1;
     }
 
-    /*
-     * If we've sent a close_notify but not yet received one back then ditch
-     * anything we read.
-     */
     if ((s->shutdown & SSL_SENT_SHUTDOWN) != 0) {
-        /*
-         * In TLSv1.3 this could get problematic if we receive a KeyUpdate
-         * message after we sent a close_notify because we're about to ditch it,
-         * so we won't be able to read a close_notify sent afterwards! We don't
-         * support that.
-         */
-        SSL3_RECORD_set_length(rr, 0);
-        SSL3_RECORD_set_read(rr);
-
         if (SSL3_RECORD_get_type(rr) == SSL3_RT_HANDSHAKE) {
             BIO *rbio;
+            /*
+             * We need to check the message type of the received data. We're not
+             * "in init" otherwise we won't have got here - so this is a new
+             * handshake message. However it's theoretically possible for us to
+             * have received part of the message header before us sending
+             * close_notify, and for us to be receiving the remainder of it now.
+             */
+            unsigned char mt = s->rlayer.handshake_fragment_len > 0
+                               ? s->rlayer.handshake_fragment[0]
+                               : *SSL3_RECORD_get_data(rr);
 
-            if ((s->mode & SSL_MODE_AUTO_RETRY) != 0)
-                goto start;
+            /*
+             * We ignore any handshake messages sent to us unless they are
+             * TLSv1.3 NewSessionTicket or KeyUpdate, in which case we want to
+             * process them. For all other handshake messages we can't do
+             * anything reasonable with them because we are unable to write any
+             * response due to having already sent close_notify.
+             */
+            if (!SSL_IS_TLS13(s)
+                       /* Shouldn't ever be 0 but lets be safe */
+                    || SSL3_RECORD_get_length(rr) == 0
+                    || (mt != SSL3_MT_NEWSESSION_TICKET
+                        && mt != SSL3_MT_KEY_UPDATE)) {
+                SSL3_RECORD_set_length(rr, 0);
+                SSL3_RECORD_set_read(rr);
```

Comment:
I would argue that instead of just throwing it away, we should go to the error state. An application should not bring us in this state ever.

Reply:
In the other PR you quoted this about client authentication: `between sending the CertificateRequest and receiving a response` The client should still get authenticated, the server is not going to request client authentication if it's happy to just not authenticate. If the client already send a shutdown, and then is being asked to authenticate, the server is going to fail, I see no point in the client not also failing.

Reply:
> An application should not bring us in this state ever.

Yes it can. As I previously mentioned it is perfectly possible for the peer to have not read our close_notify yet even though we have sent it.

> The client should still get authenticated, the server is not going to request client authentication if it's happy to just not authenticate.

Refusal to authenticate is not necessary fatal to an application. It may just mean that access to certain things will not be possible. In fact the RFC allows clients to refuse to authenticate.

> If the client already send a shutdown, and then is being asked to authenticate, the server is going to fail, I see no point in the client not also failing.

Those are application level decisions which are not up to us to decide.